### PR TITLE
feat: vscode extension

### PR DESCRIPTION
**Pull request by Slowlife01**
image
## Create Command
image
image

image
image

> **Note**
This is when you exit midway

image

## Modify Command
image
image
image
image

See [README](https://github.com/Slowlife01/pmd/tree/vscode/apps/extension#readme) for additional info

### CODE REVIEW

**Comment by Slowlife01** [comment]
Might add translate command later

**Comment by Slowlife01** [comment]
image
Would be nice if someone could fix this, if I exclude these files, require.resolve() won't be able to find the modules
image

Also switched to yarn due to (Could switch to back pnpm if the issue above is resolved) https://github.com/microsoft/vscode-vsce/issues/421

[Working build](https://github.com/PreMiD/pmd/blob/5aeff76aead699786b80f58a4790779e0d4eec44/apps/extension/vspmd-1.0.0.vsix?raw=true)

**Comment by Slowlife01** [comment]
Should be fine now, please test it
> **Note**
Run  `vsce package --no-dependencies` to build the Extension
Run `code --install-extension {path}` to install the Extension

**Review by Bas950 — COMMENTED** [review]
Maybe move apps/extensions and apps/cli to the main folder?

File: cli/LICENSE [hunk 1]
```diff
@@ -0,0 +1 @@
+../../LICENSE
```
Comment (Bas950) [inline cli/LICENSE]: this ../ probs aint correct anymore

**Review by Bas950 — CHANGES_REQUESTED** [review]

**Review by Bas950 — COMMENTED** [review]
something to discuss with @Timeraa, should we .gitignore the bin folder or not.

**Comment by Timeraa** [comment]
Sorry for the delay, why aren't you reusing classes like the ModuleManager from the cli or things like fetchSchema? I noticed that you are reusing other functions from it but not everything.

**Comment by Bas950** [comment]
image
Its impossible to see what you are inputting after you have touched the input field, a URL for image, thumbnail, website url? what url?

**Comment by Bas950** [comment]
image
is it me or does enter not do anything now?

**Comment by Slowlife01** [comment]
> image
> is it me or does enter not do anything now?

I think the validation failed and it doesn't show the error message

**Comment by Bas950** [comment]
image

**Comment by Timeraa** [comment]
Hey @Slowlife01,

Sorry this took so long to get reviewed, lot's of things came into my way.

Can you resolve the conflicts on this PR? I also have some improvements you could add to the Extension which you can add now or after the initial release:

- When modifying a Presence, you could automatically open the Presence's files (presence.ts &/ metadata.json) or show a popup like you already do when you created the Presence 
image
- When you freshly cloned the repository, you could automatically install the dependencies when you run modify/create a Presence
- When you are already modifying a Presence and run the command again it should stop the old running compiler as it's no longer needed (imo don't be afraid to give your thoughts on it)
- Perhaps you can add Open metadata.json to this popup as well
image

I have also found some errors in the Extension:
- If you run the compiler (modify) and then kill the terminal of it
image
You get this error: 
image


Also once I opened the package.json DeepScan showed me this, maybe take a look at that? 
image

File: README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (Timeraa) [inline README.md]: Just an empty line add?

File: cli/LICENSE [hunk 3]
```diff
@@ -0,0 +1 @@
+../LICENSE
```
Comment (Timeraa) [inline cli/LICENSE]: I am not sure but can you do it like that?

File: cli/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    "declaration": true,
+    "rootDir": "./src",
+    "outDir": "./bin",
+    "types": ["../extension/types"],
+    "paths": {
+      "@ext/*": ["../extension/types/*"]
+    }
+  },
+  "include": ["./src/**/*"],
+  "exclude": ["templete"]
```
Comment (Timeraa) [inline cli/tsconfig.json]: typo `templete` > `template`

File: cli/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  "compilerOptions": {
+    "declaration": true,
+    "rootDir": "./src",
+    "outDir": "./bin",
+    "types": ["../extension/types"],
+    "paths": {
+      "@ext/*": ["../extension/types/*"]
+    }
+  },
+  "include": ["./src/**/*"],
```
Comment (Timeraa) [inline cli/tsconfig.json]: Pretty sure this can be removed as `rootDir` is src

File: extension/.vscode/launch.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+      "name": "Run Extension",
+      "type": "extensionHost",
+      "request": "launch",
+      "runtimeExecutable": "${execPath}",
+      "args": ["--extensionDevelopmentPath=${workspaceFolder}"],
+      "outFiles": ["${workspaceFolder}/dist/**/*.js"],
+      "preLaunchTask": "npm: debug"
+    }
+  ]
+}
```
Comment (Timeraa) [inline extension/.vscode/launch.json]: File not needed since you have it in root

File: extension/.vscode/tasks.json [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+          "reveal": "never"
+        },
+        "group": {
+          "kind": "build",
+          "isDefault": true
+        }
+      }
+    ]
+  }
+
```
Comment (Timeraa) [inline extension/.vscode/tasks.json]: ditto

File: extension/LICENSE [hunk 8]
```diff
@@ -0,0 +1 @@
+../LICENSE
```
Comment (Timeraa) [inline extension/LICENSE]: Does it work like that?

File: extension/.vscodeignore [hunk 9]
```diff
@@ -0,0 +1,10 @@
+.vscode/
+src/
+
+*.ts
+
+dist/*.ts
+dist/*.js.map
+
+tsconfig.json
+yarn.lock
```
Comment (Timeraa) [inline extension/.vscodeignore]: shouldn't this be ´pnpm-lock.yaml` instead of `yarn.lock`?

File: extension/package.json [hunk 10]
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "vspmd",
```
Comment (Timeraa) [inline extension/package.json]: I think this should be `PreMiD pmd` as it's project bound to not cause any conflicts?

**Review by Timeraa — CHANGES_REQUESTED** [review]

**Comment by Slowlife01** [comment]
>When you are already modifying a Presence and run the command again it should stop the old running compiler as it's no longer needed (imo don't be afraid to give your thoughts on it)

Bas wanted it to be able to compile multiple Presences at the same time

File: README.md (hunk 2 above)
Comment (Slowlife01) [reply README.md]: *Will remove it for now since I'm not really good at writing one*

**Review by Timeraa — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**